src/algebra/homology/homology.lean

```diff
@@ -4,143 +4,179 @@ Released under Apache 2.0 license as described in the file LICENSE.
 variables [has_images V] [has_image_maps V]
 
-/-- A morphism of cochain complexes induces a morphism on the images of the differentials in every
+/-- A morphism of complexes induces a morphism on the images of the differentials in every
     degree. -/
-abbreviation image_map {C C' : cochain_complex V} (f : C ⟶ C') (i : ℤ) :
+abbreviation image_map {C C' : homological_complex V b} (f : C ⟶ C') (i : β) :
   image (C.d i) ⟶ image (C'.d i) :=
-image.map (arrow.hom_mk' (cochain_complex.comm_at f i).symm)
+image.map (arrow.hom_mk' (homological_complex.comm_at f i).symm)
 
 @[simp]
-lemma image_map_ι {C C' : cochain_complex V} (f : C ⟶ C') (i : ℤ) :
-  image_map f i ≫ image.ι (C'.d i) = image.ι (C.d i) ≫ f.f (i + 1) :=
-image.map_hom_mk'_ι (cochain_complex.comm_at f i).symm
+lemma image_map_ι {C C' : homological_complex V b} (f : C ⟶ C') (i : β) :
+  image_map f i ≫ image.ι (C'.d i) = image.ι (C.d i) ≫ f.f (i + b) :=
+image.map_hom_mk'_ι (homological_complex.comm_at f i).symm
 
-end
+end has_image_maps
 
-/-!
-At this point we assume that we have all images, and all equalizers.
-We need to assume all equalizers, not just kernels, so that
-`factor_thru_image` is an epimorphism.
--/
-variables [has_kernels V] [has_images V] [has_equalizers V]
+variables [has_images V] [has_equalizers V]
 
 /--
-The connecting morphism from the image of `d i` to the kernel of `d (i+1)`.
+The connecting morphism from the image of `d i` to the kernel of `d (i ± 1)`.
 -/
-def image_to_kernel_map (C : cochain_complex V) (i : ℤ) :
-  image (C.d i) ⟶ kernel (C.d (i+1)) :=
-kernel.lift _ (image.ι (C.d i)) $ (cancel_epi (factor_thru_image (C.d i))).1 $ by simp
+def image_to_kernel_map (C : homological_complex V b) (i : β) :
+  image (C.d i) ⟶ kernel (C.d (i+b)) :=
+category_theory.image_to_kernel_map (C.d i) (C.d (i+b)) (by simp)
 
 @[simp, reassoc]
-lemma image_to_kernel_map_condition (C : cochain_complex V) (i : ℤ) :
-  image_to_kernel_map C i ≫ kernel.ι (C.d (i + 1)) = image.ι (C.d i) :=
-by simp [image_to_kernel_map]
+lemma image_to_kernel_map_condition (C : homological_complex V b) (i : β) :
+  image_to_kernel_map C i ≫ kernel.ι (C.d (i + b)) = image.ι (C.d i) :=
+by simp [image_to_kernel_map, category_theory.image_to_kernel_map]
 
 @[reassoc]
-lemma induced_maps_commute [has_image_maps V] {C C' : cochain_complex V} (f : C ⟶ C')
-  (i : ℤ) :
-  image_to_kernel_map C i ≫ kernel_map f (i + 1) = image_map f i ≫ image_to_kernel_map C' i :=
+lemma induced_maps_commute [has_image_maps V] {C C' : homological_complex V b} (f : C ⟶ C')
+  (i : β) :
+  image_to_kernel_map C i ≫ kernel_map f (i + b) = image_map f i ≫ image_to_kernel_map C' i :=
 by { ext, simp }
 
 variables [has_cokernels V]
 
-/-- The `i`-th cohomology group of the cochain complex `C`. -/
-def cohomology (C : cochain_complex V) (i : ℤ) : V :=
-cokernel (image_to_kernel_map C (i-1))
+/-- The `i`-th homology group of the complex `C`. -/
+def homology (C : homological_complex V b) (i : β) : V :=
```

Comment:
Should this swap the arguments and become a functor?
I see that you are giving a more general functor below. So now I don't know...

Reply:
Ah, I see, the point is that functoriality for a single homology group is currently hard to access. Hrm. 

Okay. I'll rename this to `homology_group` (and swap the arguments, and use `C.homology_group` anyway) and make `homology i` into a functor to V.